### PR TITLE
osd: use unique_ptr for managing life cycles

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -9111,14 +9105,14 @@ void PrimaryLogPG::_copy_some_manifest(ObjectContextRef obc, CopyOpRef cop, uint
...
-				    flags,
-				    new C_OnFinisher(fin, osd->objecter_finishers[n]),
-				    // discover the object version if we don't know it yet
-				    sub_cop->results.user_version ? NULL : &sub_cop->results.user_version);
+
+    ceph_tid_t tid = osd->objecter->read(
+      soid.oid, oloc, op,
+      sub_cop->src.snap, NULL,
+      flags,
+      new C_OnFinisher(fin, osd->get_objecter_finisher(get_pg_shard())),
```
Comment (ronen-fr) [inline src/osd/PrimaryLogPG.cc]: might be a good idea to break this statement

File: src/osd/OSD.h [hunk 2]
```diff
@@ -515,11 +515,17 @@ class OSDService {
...
+  unsigned get_num_shards() const {
+    return m_objecter_finishers;
+  }
+  Finisher* get_objecter_finisher(int shard) {
+    return objecter_finishers[shard].get();
+  }
 
   // -- Objecter, for tiering reads/writes from/to other OSDs --
-  Objecter *objecter;
+  std::unique_ptr<Objecter> objecter;
```
Comment (cbodley) [inline src/osd/OSD.h]: can objecter just be a value member to avoid heap allocation entirely?

Comment (tchaikov) [reply src/osd/OSD.h]: yeah, actually, i thought about this. but i am wondering if we intended to avoid `#include "osdc/Objecter.h"` in `OSD.h`.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-12-09_12:39:45-rados-wip-kefu-testing-2019-12-09-1434-distro-basic-smithi/

failure tracked / fixed by
- https://tracker.ceph.com/issues/43192
- https://tracker.ceph.com/issues/42764
- https://tracker.ceph.com/issues/43070
- #32143

**Comment by tchaikov** [comment]
jenkins test make check

**Pull request merged**